sdk/communication/Azure.Communication.Common/src/EntraTokenCredential.cs

```diff
@@ -0,0 +1,191 @@
+using Azure.Core;
+using Azure.Core.Pipeline;
+
+namespace Azure.Communication
+{
+    /// <summary>
+    /// Represents a credential that exchanges an Entra token for an Azure Communication Services (ACS) token, enabling access to ACS resources.
+    /// </summary>
+    internal sealed class EntraTokenCredential : ICommunicationTokenCredential
+    {
+        private HttpPipeline _pipeline;
+        private string _resourceEndpoint;
+        private TokenCredential _tokenCredential;
+        private string[] _scopes;
+        private AccessToken _currentToken;
+        private readonly object _syncLock = new object();
+        private bool _someThreadIsExchanging;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="EntraTokenCredential"/>.
+        /// </summary>
+        /// <param name="options">The options for how the token will be fetched</param>
+        public EntraTokenCredential(EntraCommunicationTokenCredentialOptions options)
+        {
+            this._tokenCredential = options.TokenCredential;
+            this._resourceEndpoint = options.ResourceEndpoint;
+            this._scopes = options.Scopes;
+            _pipeline = CreatePipelineFromOptions(options);
+        }
+
+        private HttpPipeline CreatePipelineFromOptions(EntraCommunicationTokenCredentialOptions options)
+        {
+            var authenticationPolicy = new BearerTokenAuthenticationPolicy(options.TokenCredential, options.Scopes);
+            return HttpPipelineBuilder.Build(ClientOptions.Default, authenticationPolicy);
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            _currentToken = default;
+            _pipeline = default;
+        }
+
+        /// <summary>
+        /// Gets an <see cref="AccessToken"/>.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token for the task.</param>
+        /// <returns> Contains the access token.</returns>
+        public AccessToken GetToken(CancellationToken cancellationToken)
+        {
+            return GetTokenAsync(cancellationToken).Result;
+        }
+
+        /// <summary>
+        /// Gets an <see cref="AccessToken"/>.
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token for the task.</param>
+        /// <returns>
+        /// A task that represents the asynchronous get token operation. The value of its <see cref="ValueTask{AccessToken}.Result"/> property contains the access token.
+        /// </returns>
+        public ValueTask<AccessToken> GetTokenAsync(CancellationToken cancellationToken)
+        {
+            return GetValueAsync(cancellationToken);
+        }
+
+        private async ValueTask<AccessToken> GetValueAsync(CancellationToken cancellationToken)
+        {
+            if (IsTokenValid(_currentToken))
+                return _currentToken;
+
+            var shouldThisThreadExchange = false;
+            lock (_syncLock)
```

Comment:
Have you looked into whether ThreadSafeRefreshableAccessTokenCache could be reused?

Reply:
Great suggestion! I'll see if I can reuse that logic.

Reply:
nope, it was not, fixed.